doc/introduction/mid_circuit_measurements.rst

```diff
@@ -0,0 +1,530 @@
+
+.. code-block:: pycon
+
+    deferred_qnode = qml.defer_measurements(my_qnode)
+    pars = np.array([0.643, 0.246])
+    >>> deferred_qnode(*pars)
+    (tensor([0.90165331, 0.09834669], requires_grad=True),
+     tensor(0.09984972, requires_grad=True))
+
+The effect of deferring measurements becomes clear if we draw the ``QNode``
+before and after applying the transform:
+
+.. code-block:: pycon
+
+    >>> qml.draw(my_qnode)(*pars)
+    0: ──RY(0.64)─╭●───────RY(0.25)─┤  Probs
+    1: ───────────╰X──┤↗├──║────────┤
+                       ╚═══╩════════╡  <MCM>
+    >>> qml.draw(deferred_qnode)(*pars)
+    0: ──RY(0.64)─╭●────╭RY(0.25)─┤  Probs
+    1: ───────────╰X─╭●─│─────────┤
+    2: ──────────────╰X─╰●────────┤  <None>
+
+Mid-circuit measurements are deferred to the end of the circuit, and conditionally applied
+operations become (quantumly) controlled operations.
+``qml.defer_measurements`` can be applied as decorator equally well:
+
+.. code-block:: python
+
+    @qml.qnode(dev)
+    @qml.defer_measurements
+    def qnode(x, y):
+        (...)
+
+    @qml.defer_measurements
+    @qml.qnode(dev)
+    def qnode(x, y):
+        (...)
+
+.. note::
+
+    The deferred measurements principle requires an additional wire, or qubit, for each mid-circuit
+    measurement, limiting the number of measurements that can be used both on classical simulators
+    and quantum hardware. 
+
+.. _one_shot_transform:
+
+Dynamic one-shot sampling
+*************************
+
+Devices that natively support mid-circuit measurements (defined using
+:func:`~.pennylane.measure`) and conditional operations (defined using
+:func:`~.pennylane.cond`) can estimate dynamic circuits by executing
+them one shot at a time, sampling a dynamic circuit execution path for each
+shot.
+This technique is the default for a shot-based
+:class:`~.pennylane.QNode` that uses a device supporting mid-circuit measurements,
+as well as any :class:`~.pennylane.QNode` with the
+:func:`~.pennylane.dynamic_one_shot` quantum function transform.
+As the name suggests, this transform only works for a :class:`~.pennylane.QNode`
+executing with finite shots and it will raise an error if the device does not support
+mid-circuit measurements natively.
+
+The :func:`~.pennylane.dynamic_one_shot` transform is usually advantageous compared
+with the :func:`~.pennylane.defer_measurements` transform in the
+many-mid-circuit-measurements and few-shots limit. This is because, unlike the
+deferred measurement principle, the method does not need an additional wire for every
+mid-circuit measurement in the circuit.
+The transform can be applied to a QNode as follows:
+
+.. code-block:: python
+
+    @qml.dynamic_one_shot
+    @qml.qnode(dev)
+    def my_quantum_function(x, y):
+        (...)
+
+.. warning::
+
+    Dynamic circuits executed with shots should be differentiated with the finite difference method.
+
+.. _tree_traversal:
+
+Tree-traversal algorithm
+************************
```

Comment:
For tree traversal I think it also makes sense to have a warning about lack of jit compatibility